modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators.Sample/MustBeVariantSamples.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+
+namespace Godot.SourceGenerators.Sample;
+
+public class MustBeVariantMethods
+{
+    public MustBeVariantMethods()
+    {
+        Method<string>();
+
+        // This call must fail
+        // Method<NotVariant>();
+    }
+
+    public void Method<[MustBeVariant] T>() where T : class
+    {
+    }
+}
+
+public class MustBeVariantAnnotatedMethods
+{
+    [GenericTypeAttribute<string>()]
+    public void MethodWithAttribute()
+    {
+    }
+
+    // This method fails
+    /*
+    [GenericTypeAttribute<NotVariant>()]
+    public void MethodWithWrongAttribute()
+    {
+    }
+    */
```

Comment:
This is unfortunate, because the SourceGenerators.Sample project must build so we can't have invalid cases to test. Using an actual unit testing would help here, this is implemented by https://github.com/godotengine/godot/pull/82955.

Would you be able to rebase your PR after https://github.com/godotengine/godot/pull/82955 gets merged and add the tests for these cases? It's fine if you can't, someone else can contribute the tests as a follow-up PR.

Reply:
I'll try to do my best. Do you want me to delete the `MustBeVariantSamples.cs`, just in case the test is beyond of my knowledge or it takes me too much time, so you can merge this PR as a quick fix?

Reply:
I think it's fine to keep the sample for now, the SourceGenerators.Sample project still exists and can be an useful reference for implementing the unit test. Also, adding unit tests in this PR would be nice but shouldn't be a blocker, sorry if I made it sound that way.

Reply:
I will prepare the tests in another PR, so we can get this fixed asap.

Reply:
`4.3-dev.1` will be the last release of the year and unfortunately it's too late to include the fix, so there's no hurry if you still want to include the tests in this PR.

Reply:
That sounds good to me. Don't feel like you have to add tests for everything though, but if you are up for it, it's much appreciated.

Reply:
I've tried to add a test, but as soon as I add this code to the `MustBeVariant.GD0301.cs`:

```cs
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class GenericTypeAttribute<[MustBeVariant] T> : Attribute
{
}
```

The  generator fails with:
```sh
// MustBeVariant.GD0301.cs(75,56): error CS0698: A generic type cannot derive from 'Attribute' because it is an attribute class
DiagnosticResult.CompilerError("CS0698").WithSpan("MustBeVariant.GD0301.cs", 75, 56, 75, 65).WithArguments("System.Attribute"),
```
Generic attributes are allowed since C# 11, but I can't specify the 11 version in the `CSharpAnalyzerVerifier` class, which is the responsible of the compiler and analyzer configuration. Upgrade `Godot.SourceGenerators.csproj` to csharp 11 doesn't fix it. And do it the same for `GodotSharp.csproj` ends up with a ton of error. So, I can't (or I don't know how to) add specific tests for generic attributes with the [MustBeVariant] attribute. 

Anyway, the fix works because you can play with `MustBeVariantSamples.cs` and see it in action: ir works with all the variant types, but it fails when you uncomment the code commented as "it should fail".

Reply:
We'll eventually remove the sample project so it's important to make sure we add the cases that this PR fixes to the tests. Otherwise, we may reintroduce the issue in the future.

Changing the language version of the `Godot.SourceGenerators` or the `GodotSharp` projects would be incorrect. The version of C# used by source generators is not relevant to the version of C# that is generated. The version of C# used by the `GodotSharp` assembly is also unrelated since the generated code is added to the user project assembly.

You would need to change the language version of the "user project" or the C# project that is compiled which is the test sources. So your first guess was correct, you need to change it in the project created by `CSharpAnalyzerVerifier`.

To do this you can use the project's parse options:

```csharp
Project project =
    solution.GetProject(projectId)!
        .AddMetadataReference(Constants.GodotSharpAssembly.CreateMetadataReference())
        .WithParseOptions(new CSharpParseOptions(LanguageVersion.CSharp11));
```

I think you'll need to add a dependency to a newer version of `Microsoft.CodeAnalysis.CSharp.Workspaces` in order to get `LanguageVersion.CSharp11`.

Reply:
Thanks! It worked like a charm! :)

Reply:
No, I don't think that's necessary.